app/master/build.py

```diff
@@ -182,20 +184,32 @@ def execute_next_subjob_or_teardown_slave(self, slave):
         :type slave: Slave
         """
         try:
-            subjob = self._unstarted_subjobs.get(block=False)
-            self._logger.debug('Sending subjob {} (build {}) to slave {}.',
-                               subjob.subjob_id(), subjob.build_id(), slave.url)
-            slave.start_subjob(subjob)
+            # This lock prevents the scenario where a subjob is pulled from the queue but cannot be assigned to this
+            # slave because it is shutdown, so we put it back on the queue but in the meantime another slave enters
+            # this method, finds the subjob queue empty, and is torn down.  If that was the last 'living' slave, the
+            # build would be stuck.
+            with self._subjob_assignment_lock:
+                subjob = self._unstarted_subjobs.get(block=False)
+                self._logger.debug('Sending subjob {} (build {}) to slave {}.',
+                                   subjob.subjob_id(), subjob.build_id(), slave.url)
+                try:
+                    slave.start_subjob(subjob)
+                except ShutdownSlaveError:
+                    self._unstarted_subjobs.put(subjob)
+                    self._free_slave_executor(slave)
 
         except Empty:
-            num_executors_in_use = slave.free_executor()
-            if num_executors_in_use == 0:
-                try:
-                    self._slaves_allocated.remove(slave)
-                except ValueError:
-                    pass  # We have already deallocated this slave, no need to teardown
-                else:
-                    slave.teardown()
+            self._free_slave_executor(slave)
+
+    def _free_slave_executor(self, slave):
```

Comment:
If frees the slave executor each time, it may optionally teardown.  Not sure if I should include that in the function name.